modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/logical/ProjectScanMergeRule.java

```diff
@@ -92,53 +80,73 @@ private ProjectScanMergeRule(Config config) {
-                return inputRef;
-            }
-        }.apply(cond);
-
-        ImmutableBitSet requiredColumns = builder.build();
-
-        Mappings.TargetMapping targetMapping = Commons.mapping(requiredColumns,
-            tbl.getRowType(typeFactory).getFieldCount());
-
-        projects = new RexShuttle() {
-            @Override public RexNode visitInputRef(RexInputRef ref) {
-                return new RexLocalRef(targetMapping.getTarget(ref.getIndex()), ref.getType());
-            }
-        }.apply(projects);
+        if (requiredColumns == null) {
+            assert scanProjects == null;
+
+            ImmutableBitSet.Builder builder = ImmutableBitSet.builder();
+
+            new RexShuttle() {
+                @Override public RexNode visitInputRef(RexInputRef ref) {
+                    builder.set(ref.getIndex());
+                    return ref;
+                }
+            }.apply(projects);
+
+            new RexShuttle() {
+                @Override public RexNode visitLocalRef(RexLocalRef inputRef) {
+                    builder.set(inputRef.getIndex());
+                    return inputRef;
+                }
+            }.apply(cond);
+
+            requiredColumns = builder.build();
+
+            Mappings.TargetMapping targetMapping = Commons.mapping(requiredColumns,
+                tbl.getRowType(typeFactory).getFieldCount());
+
+            projects = new RexShuttle() {
+                @Override public RexNode visitInputRef(RexInputRef ref) {
+                    return new RexLocalRef(targetMapping.getTarget(ref.getIndex()), ref.getType());
+                }
+            }.apply(projects);
+
+            cond = new RexShuttle() {
+                @Override public RexNode visitLocalRef(RexLocalRef ref) {
+                    return new RexLocalRef(targetMapping.getTarget(ref.getIndex()), ref.getType());
+                }
+            }.apply(cond);
+        }
+        else
+            projects = RexUtils.replaceInputRefs(projects);
 
         if (RexUtils.isIdentity(projects, tbl.getRowType(typeFactory, requiredColumns), true))
             projects = null;
 
-        cond = new RexShuttle() {
-            @Override public RexNode visitLocalRef(RexLocalRef ref) {
-                return new RexLocalRef(targetMapping.getTarget(ref.getIndex()), ref.getType());
+        if (scanProjects != null) {
+            if (projects != null) {
+                // Merge projects.
+                projects = new RexShuttle() {
+                    @Override public RexNode visitLocalRef(RexLocalRef ref) {
+                        return scanProjects.get(ref.getIndex());
+                    }
+                }.apply(projects);
+
+                // Check again after merge.
+                if (RexUtils.isIdentity(projects, tbl.getRowType(typeFactory, requiredColumns), true))
+                    projects = null;
             }
-        }.apply(cond);
+            else
+                projects = scanProjects;
```

Comment:
Such transformation is not valid because it could possibly change the row type of the resulting rel. Please assume the follow case:

```
Project($0, $1, $2)
    Scan(proj=[$0, $1, $2, <some constant or constant expression like 2 * 2>])
```

So you should first merge projects and then discard the result if it is identity

Reply:
Fixed, thanks